plutus-playground/plutus-playground-server/app/Auth.hs

```diff
@@ -0,0 +1,315 @@
+import           Network.HTTP.Types          (hAccept, statusIsSuccessful)
+import           Servant                     ((:<|>) ((:<|>)), (:>), Get, Header, Headers, JSON, NoContent (NoContent),
+                                              QueryParam, ServantErr, ServerT, StdMethod (GET), ToHttpApiData, Verb,
+                                              addHeader, err401, err500, errBody, throwError)
+import           Servant.API.BrowserHeader   (BrowserHeader)
+import           Servant.Client              (BaseUrl, mkClientEnv, parseBaseUrl, runClientM)
+import           Servant.Extra               ()
+import           Web.Cookie                  (SetCookie, defaultSetCookie, parseCookies, setCookieExpires,
+                                              setCookieHttpOnly, setCookieMaxAge, setCookieName, setCookiePath,
+                                              setCookieSecure, setCookieValue)
+import qualified Web.JWT                     as JWT
+
+data AuthRole
+  = Anonymous
+  | GithubUser
+  deriving (Show, Eq, Generic, FromJSON, ToJSON)
+
+newtype AuthStatus = AuthStatus
+  { _authStatusAuthRole :: AuthRole
+  } deriving (Show, Eq, Generic, FromJSON, ToJSON)
+
+-- | https://gist.github.com/alpmestan/757094ecf9401f85c5ba367ca20b8900
+type GetRedirect headers = Verb 'GET 302 '[ JSON] (headers NoContent)
+
+-- | We split out the API here because we don't want the Github
+-- Callback to be exported for the frontend to call directly.
+type API
+   = FrontendAPI
+     :<|> CallbackAPI
+
+type FrontendAPI
+   = ("oauth" :> (BrowserHeader "Cookie" Text :> "status" :> Get '[ JSON] AuthStatus
+                  :<|> "github" :> GetRedirect (Headers '[ Header "Location" Text])))
+     :<|> ("gists" :> BrowserHeader "Cookie" Text :> Get '[ JSON] [Gist])
+
+type CallbackAPI
+   = "oauth" :> "github" :> "callback" :> QueryParam "code" OAuthCode :> GetRedirect (Headers '[ Header "Set-Cookie" SetCookie, Header "Location" Text])
+
+data GithubEndpoints = GithubEndpoints
+  { _githubEndpointsAuthLocation        :: !Request
+  , _githubEndpointsAccessTokenLocation :: !Request
+  , _githubEndpointsApiBaseUrl          :: !BaseUrl
+  , _githubEndpointsCallbackUri         :: !Text
+  }
+
+-- | Config determined by Github.
+mkGithubEndpoints :: IO GithubEndpoints
+mkGithubEndpoints = do
+  _githubEndpointsAuthLocation <-
+    parseRequest "GET https://github.com/login/oauth/authorize"
+  _githubEndpointsAccessTokenLocation <-
+    parseRequest "POST https://github.com/login/oauth/access_token"
+  _githubEndpointsApiBaseUrl <- parseBaseUrl "https://api.github.com"
+  let _githubEndpointsCallbackUri = "/api/oauth/github/callback"
+  pure GithubEndpoints {..}
+
+-- | Config supplied at runtime.
+data Config = Config
+  { _configJWTSignature       :: !JWT.Signer
+  , _configRedirectUrl        :: !Text
+  , _configRedirectEndpoint   :: !Text
+  , _configGithubClientId     :: !Text
+  , _configGithubClientSecret :: !Text
+  }
```

Comment:
can't we improve on `Text` types here?

Reply:
![1100935_1](https://user-images.githubusercontent.com/380756/51492056-5e944f80-1da8-11e9-8e2b-d14cdae62274.jpg)